const-oid/src/parser.rs

```diff
@@ -0,0 +1,162 @@
+            7 => ObjectIdentifier::new(&[
+                arcs[0], arcs[1], arcs[2], arcs[3], arcs[4], arcs[5], arcs[6],
+            ]),
+            8 => ObjectIdentifier::new(&[
+                arcs[0], arcs[1], arcs[2], arcs[3], arcs[4], arcs[5], arcs[6], arcs[7],
+            ]),
+            9 => ObjectIdentifier::new(&[
+                arcs[0], arcs[1], arcs[2], arcs[3], arcs[4], arcs[5], arcs[6], arcs[7], arcs[8],
+            ]),
+            10 => ObjectIdentifier::new(&[
+                arcs[0], arcs[1], arcs[2], arcs[3], arcs[4], arcs[5], arcs[6], arcs[7], arcs[8],
+                arcs[9],
+            ]),
+            11 => ObjectIdentifier::new(&[
+                arcs[0], arcs[1], arcs[2], arcs[3], arcs[4], arcs[5], arcs[6], arcs[7], arcs[8],
+                arcs[9], arcs[10],
+            ]),
+            12 => ObjectIdentifier::new(&[
+                arcs[0], arcs[1], arcs[2], arcs[3], arcs[4], arcs[5], arcs[6], arcs[7], arcs[8],
+                arcs[9], arcs[10], arcs[11],
+            ]),
+            _ => ObjectIdentifier::new(&[]),
+        }
+    }
+
+    /// Parse the remaining bytes
+    const fn parse_bytes(mut self, bytes: &[u8]) -> Self {
+        match bytes {
+            [] => {
+                self.cursor += 1;
+                self
+            }
+            [byte @ b'0'..=b'9', remaining @ ..] => {
+                let current_arc = self.arcs[self.cursor];
+                self.arcs[self.cursor] = current_arc * 10 + parse_ascii_digit(*byte);
+                self.parse_bytes(remaining)
+            }
+            [b'.', remaining @ ..] => {
+                const_assert!(!remaining.is_empty(), "invalid trailing '.' in OID");
+                const_assert!(
+                    self.cursor < MAX_ARCS,
+                    "maximum number of OID arcs exceeded"
+                );
+                self.cursor += 1;
+                self.parse_bytes(remaining)
+            }
+            [byte, ..] => {
+                const_assert!(
+                    matches!(byte, b'0'..=b'9' | b'.'),
+                    "invalid character in OID"
+                );
+
+                // Unreachable (checked by above `const_assert!`)
+                // Needed for match exhaustiveness and matching types
+                self
+            }
+        }
+    }
+}
+
+/// Parse a digit from an ASCII character
+// TODO(tarcieri): replace this with `byte.saturating_sub(b'0')` when MSRV 1.47+
+const fn parse_ascii_digit(char: u8) -> Arc {
+    match char {
+        b'0' => 0,
+        b'1' => 1,
+        b'2' => 2,
+        b'3' => 3,
+        b'4' => 4,
+        b'5' => 5,
+        b'6' => 6,
+        b'7' => 7,
+        b'8' => 8,
+        b'9' => 9,
+        other => {
+            const_assert!(matches!(other, b'0'..=b'9'), "invalid ASCII digit");
+            0 // Unreachable due to above `const_assert`
+        }
+    }
+}
```

Comment:
Oof, wrote this locally using `saturating_sub` only to discover that it was made `const fn` in Rust 1.47 and the current MSRV is 1.46. So we need these unfortunate hacks for now.

`const fn` is a bit bleeding edge but rapidly getting better 😅